teps/0033-tekton-feature-gates.md

```diff
@@ -145,6 +160,393 @@ alpha features. This is not a requirement for Tekton, because even in case of
+and features that are only behavioral. (The alternative
+[one feature flag per new alpha field](#one-feature-flag-per-new-alpha-field) describes
+an alternative where we use the same approach for both.)
+
+In both cases we want to control access with
+[a feature flag](https://github.com/tektoncd/pipeline/blob/master/docs/install.md#customizing-the-pipelines-controller-behavior).
+
+* For new API fields, will will gate access to all of them behind two feature flags
+  called `enable-alpha-api-fields` and `enable-beta-api-fields`.
+  ([Design details: new api fields](#new-api-fields).)
+* For changes in behavior which are not configurable via API fields in our CRDs,
+  each feature will be gated by its own unique feature flag.
+  ([Design details: changes in behavior](#changes-in-behavior).)
+
+
+### Risks and Mitigations
+
+See the pros and cons sections in the [design details](#design-details)
+below.
+
+## Design Details
+
+### New API fields
+
+We gate access to all new API fields with two new
+[feature flags](https://github.com/tektoncd/pipeline/blob/master/docs/install.md#customizing-the-pipelines-controller-behavior) called `enable-alpha-api-fields` and `enable-beta-api-fields`
+which will default to false.
+
+For example:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: feature-flags
+data:
+  enable-alpha-api-fields: "true"
+  enable-beta-api-fields: "true"
+```
+
+This allows administrators to opt into allowing their users to use alpha and beta fields.
+
+_Note that we will only need the `enable-beta-api-fields` flag once we have v1 types in Tekton,
+at the moment the highest version we have is beta, so we will initially only need
+`enable-alpha-api-fields`._
+
+Anyone submitting any Tekton CRD (e.g. Pipelines, Tasks, Runs, PipelineRuns, TaskRuns) when these are
+true will be able to use fields which are considered "alpha" and/or "beta" (depending on which fields
+are set). The API fields enabled by these fields may be changed according to our
+[API compatibility policy](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md)
+meaning that they could be dropped or changed with at least one release of warning. (This will be true
+even when the fields are added to types which are considered `beta` or `ga`/`v1`).
+
+Since the API fields will exist in the CRD with or without the flags, this will be enforced at runtime
+by the webhook admission controller: if the flag is not enabled and a user tries to use an
+alpha field, the webhook will not allow the CRD to be stored and will return an actionable error.
+(Same for beta.)
+
+Pros:
+* One flag (for each of the 2 levels of stability) makes it easy to opt into alpha features
+  as a whole (assumes a user who wants any of these features wants "cutting edge" features in general)
+* Provides a simpler testing matrix than one flag per field
+  (see [one feature flag per new alpha field](#one-feature-flag-per-new-alpha-field)
+  alternative for details)
+
+Cons:
```

Comment:
To fully asses pros and cons we might to get deeper into the detailed design of this approach.
Are fields always added to the latest available API version?
How do make sure APIs are "round-trippable" ?
How does an API field move from `alpha` to `beta` to `v1` exactly? 

Can a deployment have alpha flags enabled and beta flags disabled?

It may be worth mentioning that folks could use other tools to prevent access to specific features if they need to, we could provide a template admission webhook for that or some documentation.

Reply:
> Are fields always added to the latest available API version?

Yes

> How do make sure APIs are "round-trippable" ?

As we do now, but that's one reason to be able to add field to the latest available API version, so that we don't need to think about round-trippable APIs when `v1alpha1` and `v1beta1` are gone

> How does an API field move from alpha to beta to v1 exactly?

I am not fully sure I understand the question 🤔 

> Can a deployment have alpha flags enabled and beta flags disabled?

I think we shouldn't, as you commented above.